styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -428,26 +429,39 @@ boolean shouldDeletePodWithoutRunState(WorkflowInstance workflowInstance, Pod po
     }
   }
 
-  private boolean shouldDeleteNonTerminatedPodWithoutRunState(WorkflowInstance workflowInstance, String name) {
+  private PodDeletionDecision shouldDeleteNonTerminatedPodWithoutRunState(WorkflowInstance workflowInstance, String name) {
     // Fetch the pod here to avoid acting on stale information
     var podOpt = client.getPod(name);
     if (podOpt.isEmpty()) {
       // The pod is gone, nothing left to do here
-      return false;
+      return PodDeletionDecision.DO_NOT_DELETE;
     }
     var pod = podOpt.orElseThrow();
     // if not terminated, delete directly
     if (!isTerminated(pod)) {
-      return shouldDeletePod(workflowInstance, pod, "No RunState, not terminated");
+      return PodDeletionDecision.newBuilder().delete(shouldDeletePod(workflowInstance, pod, "No RunState, not terminated"))
+          .build();
     }
-    return false;
+    return PodDeletionDecision.DO_NOT_DELETE;
   }
 
-  private boolean shouldDeletePodIfNonDeletePeriodExpired(WorkflowInstance workflowInstance, Pod pod) {
+  private PodDeletionDecision shouldDeletePodIfNonDeletePeriodExpired(WorkflowInstance workflowInstance, Pod pod) {
     // if terminated and after graceful period, delete the pod
     // otherwise wait until next polling happens
-    return getMainContainerStatus(pod).map(this::isNonDeletePeriodExpired).orElse(false) &&
-           shouldDeletePod(workflowInstance, pod, "Terminated and expired");
+    return getMainContainerStatus(pod)
+        .map(containerStatus -> {
+          var timeToDeletePossiblyForce = isTimeToDeletePossiblyForce(containerStatus);
+          if (timeToDeletePossiblyForce.delete()) {
+            var reason = timeToDeletePossiblyForce.force()
+                         ? "Terminated, expired, and intolerant"
+                         : "Terminated and expired";
+            return timeToDeletePossiblyForce.builder()
+                .delete(shouldDeletePod(workflowInstance, pod, reason))
```

Comment:
this shouldDelete logic seems a bit legacy. I wonder if anyone has used it anytime recently.

Reply:
It's a useful feature that comes handy when we need to keep pods around for debugging.